File: CS/DragAndDropBetweenGridAndTreeList/Managers/GridManager.cs
// Developer Express Code Central Example:
// How to implement drag-and-drop between Grid and TreeList
// 
// This example demonstrates how to create helper classes for implementing
// drag-and-drop between Grid and TreeList. For more information please refer
// to:
// http://www.devexpress.com/scid=A2343
// http://www.devexpress.com/scid=K18056
// http://www.devexpress.com/scid=A1444
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=E3375

using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraGrid;
using System.Data;
using DevExpress.XtraTreeList.Nodes;
using System.Windows.Forms;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using System.Drawing;
using DevExpress.XtraTreeList;

namespace DragAndDropBetweenGridAndTreeList
{
    class GridManager
    {
        readonly GridControl _gridControl;
        GridHitInfo _dragStartHitInfo;
        Cursor _dragRowCursor;
        readonly DragGridImageHelper _imageHelper;

        public GridManager(GridControl grid)
        {
            _gridControl = grid;
            SetUpGrid(grid);
            _imageHelper = new DragGridImageHelper(grid.FocusedView as GridView);
        }

        public void SetUpGrid(GridControl grid)
        {
            grid.AllowDrop = true;
            grid.DragOver += grid_DragOver;
            grid.DragDrop += grid_DragDrop;
            grid.DragLeave += grid_DragLeave;
            grid.Paint += grid_Paint;
            grid.GiveFeedback += grid_GiveFeedback;
            GridView view = grid.MainView as GridView;
            view.OptionsBehavior.Editable = false;
            view.MouseMove += view_MouseMove;
            view.MouseDown += view_MouseDown;
        }

        private void view_MouseDown(object sender, MouseEventArgs e)
        {
            GridView view = sender as GridView;
            _dragStartHitInfo = null;
            GridHitInfo hitInfo = view.CalcHitInfo(new Point(e.X, e.Y));
            if (Control.ModifierKeys != Keys.None)
                return;
            if (e.Button == MouseButtons.Left && hitInfo.RowHandle >= 0)
                _dragStartHitInfo = hitInfo;
        }

        private void view_MouseMove(object sender, MouseEventArgs e)
        {
            GridView view = sender as GridView;
            if (e.Button == MouseButtons.Left && _dragStartHitInfo != null)
            {
                Size dragSize = SystemInformation.DragSize;
                Rectangle dragRect = new Rectangle(new Point(_dragStartHitInfo.HitPoint.X - dragSize.Width / 2, _dragStartHitInfo.HitPoint.Y - dragSize.Height / 2), dragSize);
                if (!dragRect.Contains(new Point(e.X, e.Y)))
                {
                    _dragRowCursor = _imageHelper.GetDragCursor(_dragStartHitInfo.RowHandle, e.Location);
                    DataRow row = view.GetDataRow(_dragStartHitInfo.RowHandle);
                    view.GridControl.DoDragDrop(row, DragDropEffects.Move);
                    _dragStartHitInfo = null;
                    DevExpress.Utils.DXMouseEventArgs.GetMouseArgs(e).Handled = true;
                }
            }
        }

        int dropTargetRowHandle = -1;
        int DropTargetRowHandle
        {
            get
            {
                return dropTargetRowHandle;
            }
            set
            {
                dropTargetRowHandle = value;
                _gridControl.Invalidate();
            }
        }

        private void grid_DragDrop(object sender, DragEventArgs e)
        {
            GridControl grid = sender as GridControl;
            DataTable table = grid.DataSource as DataTable;
            TreeListNode node = e.Data.GetData(typeof(TreeListNode)) as TreeListNode;
            if (node.RootNode == node)
                return;
            if (node != null && table != null)
            {
                List<object> itemArray = new List<object>();
                foreach (DataColumn column in table.Columns)
                {
                    itemArray.Add(node.GetValue(column));
                }
                object[] tmp = new object[itemArray.Count];
                itemArray.CopyTo(tmp);
                DataRow row = table.NewRow();
                row.ItemArray = tmp;
                table.Rows.InsertAt(row, DropTargetRowHandle);
                TreeList treeList = node.TreeList;
                treeList.DeleteNode(node);
            }
            DropTargetRowHandle = -1;
        }

        private void grid_DragOver(object sender, DragEventArgs e)
        {
            GridControl grid = (GridControl)sender;
            Point pt = new Point(e.X, e.Y);
            pt = grid.PointToClient(pt);
            GridView view = grid.GetViewAt(pt) as GridView;
            if (view == null)
                return;
            GridHitInfo hitInfo = view.CalcHitInfo(pt);
            if (hitInfo.RowHandle == GridControl.InvalidRowHandle)
                DropTargetRowHandle = view.DataRowCount;
            else
                DropTargetRowHandle = hitInfo.RowHandle;
            TreeListNode node = e.Data.GetData(typeof(TreeListNode)) as TreeListNode;
            if (DropTargetRowHandle >= 0 && e.Data.GetDataPresent(typeof(TreeListNode)) && node.RootNode != node)
                e.Effect = DragDropEffects.Move;
            else
                e.Effect = DragDropEffects.None;
        }

        private void grid_DragLeave(object sender, EventArgs e)
        {
            DropTargetRowHandle = -1;
        }

        private void grid_Paint(object sender, PaintEventArgs e)
        {
            if (DropTargetRowHandle < 0)
                return;
            GridControl grid = (GridControl)sender;
            GridView view = (GridView)grid.MainView;
            bool isBottomLine = DropTargetRowHandle == view.DataRowCount;
            GridViewInfo viewInfo = view.GetViewInfo() as GridViewInfo;
            GridRowInfo rowInfo = viewInfo.GetGridRowInfo(isBottomLine ? DropTargetRowHandle - 1 : DropTargetRowHandle);
            if (rowInfo == null)
                return;
            Point p1, p2;
            if (isBottomLine)
            {
                p1 = new Point(rowInfo.Bounds.Left, rowInfo.Bounds.Bottom - 1);
                p2 = new Point(rowInfo.Bounds.Right, rowInfo.Bounds.Bottom - 1);
            }
            else
            {
                p1 = new Point(rowInfo.Bounds.Left, rowInfo.Bounds.Top - 1);
                p2 = new Point(rowInfo.Bounds.Right, rowInfo.Bounds.Top - 1);
            }
            e.Graphics.DrawLine(Pens.Blue, p1, p2);
        }

        void grid_GiveFeedback(object sender, GiveFeedbackEventArgs e)
        {
            if (_dragStartHitInfo != null)
            {
                e.UseDefaultCursors = false;
                Cursor.Current = _dragRowCursor;
            }
        }
    }
}

File: VB/DragAndDropBetweenGridAndTreeList/Managers/GridManager.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraGrid
Imports System.Data
Imports DevExpress.XtraTreeList.Nodes
Imports System.Windows.Forms
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports System.Drawing
Imports DevExpress.XtraTreeList

Namespace DragAndDropBetweenGridAndTreeList
	Friend Class GridManager
		Private ReadOnly _gridControl As GridControl
		Private _dragStartHitInfo As GridHitInfo
		Private _dragRowCursor As Cursor
		Private ReadOnly _imageHelper As DragGridImageHelper

		Public Sub New(ByVal grid As GridControl)
			_gridControl = grid
			SetUpGrid(grid)
			_imageHelper = New DragGridImageHelper(TryCast(grid.FocusedView, GridView))
		End Sub

		Public Sub SetUpGrid(ByVal grid As GridControl)
			grid.AllowDrop = True
			AddHandler grid.DragOver, AddressOf grid_DragOver
			AddHandler grid.DragDrop, AddressOf grid_DragDrop
			AddHandler grid.DragLeave, AddressOf grid_DragLeave
			AddHandler grid.Paint, AddressOf grid_Paint
			AddHandler grid.GiveFeedback, AddressOf grid_GiveFeedback
			Dim view As GridView = TryCast(grid.MainView, GridView)
			view.OptionsBehavior.Editable = False
			AddHandler view.MouseMove, AddressOf view_MouseMove
			AddHandler view.MouseDown, AddressOf view_MouseDown
		End Sub

		Private Sub view_MouseDown(ByVal sender As Object, ByVal e As MouseEventArgs)
			Dim view As GridView = TryCast(sender, GridView)
			_dragStartHitInfo = Nothing
			Dim hitInfo As GridHitInfo = view.CalcHitInfo(New Point(e.X, e.Y))
			If Control.ModifierKeys <> Keys.None Then
				Return
			End If
			If e.Button = MouseButtons.Left AndAlso hitInfo.RowHandle >= 0 Then
				_dragStartHitInfo = hitInfo
			End If
		End Sub

		Private Sub view_MouseMove(ByVal sender As Object, ByVal e As MouseEventArgs)
			Dim view As GridView = TryCast(sender, GridView)
			If e.Button = MouseButtons.Left AndAlso _dragStartHitInfo IsNot Nothing Then
				Dim dragSize As Size = SystemInformation.DragSize
				Dim dragRect As New Rectangle(New Point(_dragStartHitInfo.HitPoint.X - dragSize.Width \ 2, _dragStartHitInfo.HitPoint.Y - dragSize.Height \ 2), dragSize)
				If (Not dragRect.Contains(New Point(e.X, e.Y))) Then
					_dragRowCursor = _imageHelper.GetDragCursor(_dragStartHitInfo.RowHandle, e.Location)
					Dim row As DataRow = view.GetDataRow(_dragStartHitInfo.RowHandle)
					view.GridControl.DoDragDrop(row, DragDropEffects.Move)
					_dragStartHitInfo = Nothing
					DevExpress.Utils.DXMouseEventArgs.GetMouseArgs(e).Handled = True
				End If
			End If
		End Sub

		Private dropTargetRowHandle_Renamed As Integer = -1
		Private Property DropTargetRowHandle() As Integer
			Get
				Return dropTargetRowHandle_Renamed
			End Get
			Set(ByVal value As Integer)
				dropTargetRowHandle_Renamed = value
				_gridControl.Invalidate()
			End Set
		End Property

		Private Sub grid_DragDrop(ByVal sender As Object, ByVal e As DragEventArgs)
			Dim grid As GridControl = TryCast(sender, GridControl)
			Dim table As DataTable = TryCast(grid.DataSource, DataTable)
			Dim node As TreeListNode = TryCast(e.Data.GetData(GetType(TreeListNode)), TreeListNode)
			If node.RootNode Is node Then
				Return
			End If
			If node IsNot Nothing AndAlso table IsNot Nothing Then
				Dim itemArray As New List(Of Object)()
				For Each column As DataColumn In table.Columns
					itemArray.Add(node.GetValue(column))
				Next column
				Dim tmp(itemArray.Count - 1) As Object
				itemArray.CopyTo(tmp)
				Dim row As DataRow = table.NewRow()
				row.ItemArray = tmp
				table.Rows.InsertAt(row, DropTargetRowHandle)
				Dim treeList As TreeList = node.TreeList
				treeList.DeleteNode(node)
			End If
			DropTargetRowHandle = -1
		End Sub

		Private Sub grid_DragOver(ByVal sender As Object, ByVal e As DragEventArgs)
			Dim grid As GridControl = CType(sender, GridControl)
			Dim pt As New Point(e.X, e.Y)
			pt = grid.PointToClient(pt)
			Dim view As GridView = TryCast(grid.GetViewAt(pt), GridView)
			If view Is Nothing Then
				Return
			End If
			Dim hitInfo As GridHitInfo = view.CalcHitInfo(pt)
			If hitInfo.RowHandle = GridControl.InvalidRowHandle Then
				DropTargetRowHandle = view.DataRowCount
			Else
				DropTargetRowHandle = hitInfo.RowHandle
			End If
			Dim node As TreeListNode = TryCast(e.Data.GetData(GetType(TreeListNode)), TreeListNode)
			If DropTargetRowHandle >= 0 AndAlso e.Data.GetDataPresent(GetType(TreeListNode)) AndAlso node.RootNode IsNot node Then
				e.Effect = DragDropEffects.Move
			Else
				e.Effect = DragDropEffects.None
			End If
		End Sub

		Private Sub grid_DragLeave(ByVal sender As Object, ByVal e As EventArgs)
			DropTargetRowHandle = -1
		End Sub

		Private Sub grid_Paint(ByVal sender As Object, ByVal e As PaintEventArgs)
			If DropTargetRowHandle < 0 Then
				Return
			End If
			Dim grid As GridControl = CType(sender, GridControl)
			Dim view As GridView = CType(grid.MainView, GridView)
			Dim isBottomLine As Boolean = DropTargetRowHandle = view.DataRowCount
			Dim viewInfo As GridViewInfo = TryCast(view.GetViewInfo(), GridViewInfo)
			Dim rowInfo As GridRowInfo = viewInfo.GetGridRowInfo(If(isBottomLine, DropTargetRowHandle - 1, DropTargetRowHandle))
			If rowInfo Is Nothing Then
				Return
			End If
			Dim p1, p2 As Point
			If isBottomLine Then
				p1 = New Point(rowInfo.Bounds.Left, rowInfo.Bounds.Bottom - 1)
				p2 = New Point(rowInfo.Bounds.Right, rowInfo.Bounds.Bottom - 1)
			Else
				p1 = New Point(rowInfo.Bounds.Left, rowInfo.Bounds.Top - 1)
				p2 = New Point(rowInfo.Bounds.Right, rowInfo.Bounds.Top - 1)
			End If
			e.Graphics.DrawLine(Pens.Blue, p1, p2)
		End Sub

		Private Sub grid_GiveFeedback(ByVal sender As Object, ByVal e As GiveFeedbackEventArgs)
			If _dragStartHitInfo IsNot Nothing Then
				e.UseDefaultCursors = False
				Cursor.Current = _dragRowCursor
			End If
		End Sub
	End Class
End Namespace
